docs/profiling.md

```diff
@@ -0,0 +1,92 @@
+return successfully. During the profile period, it also invokes [pmu-tools](https://github.com/andikleen/pmu-tools) 
+to profile user-defined counters followed by [TopDown method](https://ieeexplore.ieee.org/document/6844459). 
+After the measurement finishes, completed RPS per core, average execution time and the average 
+counters will save in the `profile.csv`.
+
+## Runtime Arguments
+```
+General:
+-warmUpTime   FLOAT  The warm up time before profiling in seconds
+-profileTime  FLOAT  The profiling time in seconds
+-coolDownTime FLOAT  The cool down time after profiling in seconds
+-loadStep     FLOAT  The percentage of target RPS the benchmark loads at every step
+-funcNames    STR    Names of the functions to benchmark, separated by comma
+-bindSocket   BOOL   Bind all VMs to socket 1 and profile one physical core only
+                     (only compatible with a 2x 16-core machine or above)
+
+TestProfileSingleConfiguration:
+-vm           INT    The number of VMs
+-rps          INT    The target requests per second
+
+TestProfileIncrementConfiguration:
+-vmIncrStep   INT    The increment VM number
+-maxVMNum     INT    The maximum VM number
+
+Profiler:
+-l            INT    Profile level
+-I            UINT   Print count deltas every N milliseconds
+-nodes        STR    Include or exclude nodes (with + to add, -|^ to remove, 
+                     comma separated list, wildcards allowed, 
+                     add * to include all children/siblings, 
+                     add /level to specify highest level node to match, 
+                     add ^ to match related siblings and metrics, 
+                     start with ! to only include specified nodes)
+```
+
+## Pre-requisites
+Assume you are at the root of this repository, run `scripts/install_pmutool.sh` to install
+essential tools for profiling and binding.
+
+## Quick-start guide
+Function `TestProfileSingleConfiguration` is for collecting counters from a fixed number of VMs and RPS setting 
+during profiling period.
+
+Profile a single VM with `helloworld` image and 20 RPS at TopDown level 1 :
+```
+sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileSingleConfiguration -args -funcNames helloworld -vm 1 -rps 20 -l 1
+```
+Bottleneck counters will be printed out with their value at each step during 
+execution, such as:
+```
+...
+INFO[] Current RPS: 1200
+INFO[] Bottleneck Backend_Bound with value 75.695000
+...
+```
+    
+To study microarchitectural bottlenecks in more detail, profile the same configuration with sub-level counters. 
+For example, the bottleneck is in the backend bound at level 1 and now we want to profile level 2 of the backend bound:
+```
+sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileSingleConfiguration -args -funcNames helloworld -vm 1 -rps 20 -nodes '!+Backend_Bound*/2,+MUX'
+```
+
+Function `TestProfileIncrementConfiguration` will load the pre-stored number of RPS and 
```

Comment:
this paragraph is unclear. please assume unexperienced user and explain step-by-step.

Reply:
Done